ops/_tracing/export.py

```diff
@@ -0,0 +1,248 @@
+
+# Trace `urllib` usage when talking to Pebble
+URLLibInstrumentor().instrument()
+
+# NOTE: nominally int, although float would work just as well in practice
+EXPORTER_TIMEOUT: int = 1  # seconds
+"""How much to give OTLP span exporter has to push traces to the backend."""
+
+SENDOUT_FACTOR: int = 2
+"""How many buffered chunks to send out for each incoming chunk."""
+
+BUFFER_FILE: str = '.tracing-data.db'
+"""Name of the file whither data is buffered, located next to .unit-state.db."""
+
+
+logger = logging.getLogger(__name__)
+_exporter: ProxySpanExporter | None = None
+
+
+# NOTE: OTEL SDK suppresses errors while exporting data
+# FIXME: decide if we need to remove this before going to prod
+logger.addHandler(logging.StreamHandler())
+
+
+class ProxySpanExporter(SpanExporter):
+    settings: tuple[str | None, str | None] = (None, None)
+    cache: dict[str | None, ssl.SSLContext]
+
+    def __init__(self, buffer_path: Path | str):
+        self.buffer = ops._tracing.buffer.Buffer(buffer_path)
+        self.lock = threading.Lock()
+        self.cache = {}
+
+    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
+        """Export a batch of telemetry data.
+
+        Note: to avoid data loops or recursion, this function cannot be instrumented.
+        """
+        try:
+            with suppress_juju_log_handler():
+                # Note:
+                # this is called in a helper thread, which is daemonic,
+                # the MainThread will wait at most 10s for this thread.
+                # Margins:
+                # - 1s safety margin
+                # - 1s for buffered data time overhang
+                # - 2s for live data
+                deadline = time.monotonic() + 6
+
+                assert spans  # the BatchSpanProcessor won't call us if there's no data
+                # TODO:  this will change in the JSON experiment
+                # __import__("pdb").set_trace()
+                # FIXME can't use stock exporter, must DIY
+
+                rv = self.buffer.pump((otlp_json.encode_spans(spans), otlp_json.CONTENT_TYPE))
+                assert rv
+                self.do_export(*rv)
+
+                for _ in range(SENDOUT_FACTOR - 1):
+                    if time.monotonic() > deadline:
+                        break
+                    if not (rv := self.buffer.pump()):
+                        break
+                    self.do_export(*rv)
+
+                return SpanExportResult.SUCCESS
+        except Exception:
+            logger.exception('export')
+            raise
+
+    def ssl_context(self, ca: str | None) -> ssl.SSLContext:
+        if context := self.cache.get(ca):
+            return context
+        context = self._ssl_context(ca)
+        self.cache.clear()
+        self.cache[ca] = context
+        return context
+
+    def _ssl_context(self, ca: str | None) -> ssl.SSLContext:
+        # FIXME: What should our protocol range be?
```

Comment:
*Maybe* 1.2+, ideally 1.3+ only.

Reply:
https://matrix.to/#/!nHXpRkcSNJHlHdUGbQ:ubuntu.com/$ymSYuWFl_KWT83WayQiF4yFQ_yVlhrLbpa_7Ns6z3zI?via=ubuntu.com&via=matrix.org for COS

I'll need to figure out a way to require 1.3 in the client that works across py 3.8 ~ 3.12.

My understanding is that newer Pythons already do that, and that's why specific protocol selection is deprecated (since 3.10?). Perhaps only 3.8 needs a kludge.

Let's say I write some code that should do the right thing... how do I validate that it does?